File: exercise-2022/simulation-2016/10.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
using namespace std;

vector<int> post, mid;
unordered_map<int, int> ind;
struct Node { int val; Node *left, *right; };

Node *build(int l, int r, int rid)
{
    if (l > r) return nullptr;
    Node *node = new Node();
    node->val = post[rid];
    int index = ind[node->val];
    node->left = build(l, index-1, rid-(r-index+1));
    node->right = build(index+1, r, rid-1);
    return node;
}

void travel(Node *root)
{
    vector<int> ans;
    queue<Node *> q;
    if (root) q.push(root);
    while (q.size())
    {
        Node *node = q.front();
        q.pop();
        ans.push_back(node->val);
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    if (ans.size()) printf("%d", ans[0]);
    for (int i=1; i<ans.size(); i++)
        printf(" %d", ans[i]);
}

int main()
{
    int n;
    cin >> n;
    for (int i=0; i<n; i++)
    {
        int num;
        cin >> num;
        post.push_back(num);
    }
    for (int i=0; i<n; i++)
    {
        int num;
        cin >> num;
        mid.push_back(num);
    }
    for (int i=0; i<n; i++)
        ind[mid[i]] = i;
    travel(build(0, n-1, n-1));
    return 0;
}

File: exercise-2022/simulation-2016/Python/10.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
def build(l, r, t):
    if l > r:
        return None
    node = Node(post[t])
    index = ind[node.val]
    node.left = build(l, index-1, t-(r-index+1))
    node.right = build(index+1, r, t-1)
    return node
def travel(root):
    if root == None:
        return
    ans = []
    q = [root]
    while len(q):
        node = q.pop(0)
        ans.append(node.val)
        if node.left != None:
            q.append(node.left)
        if node.right != None:
            q.append(node.right)
    print(*ans)
n = int(input())
post = input().split(' ')
mid = input().split(' ')
ind = {}
for i, j in enumerate(mid):
    ind[j] = i
travel(build(0, n-1, n-1))